core/src/main/java/tc/oc/pgm/loot/LootableMatchModule.java

```diff
@@ -0,0 +1,337 @@
+    } else if (holder instanceof Player) {
+      // This happens with crafting inventories, and possibly other transient inventory types
+      // Pretty sure we never want to fill an inventory held by the player
+      return null;
+    } else if (holder instanceof Entity) {
+      return filter -> !filter.query(new EntityQuery((Entity) holder)).isDenied();
+    } else {
+      // If we're not sure what it is, don't fill it
+      return null;
+    }
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
+  public void onInventoryOpen(InventoryOpenEvent event) {
+    final MatchPlayer opener = this.match.getParticipant(event.getActor());
+    if (opener == null) return;
+
+    final Inventory inventory = event.getInventory();
+    final Predicate<Filter> passesFilter = passesFilter(inventory.getHolder());
+    if (passesFilter == null) return;
+
+    logger.fine(
+        () -> opener.getName() + " opened a " + inventory.getHolder().getClass().getSimpleName());
+
+    // Find all Fillers that apply to the holder of the opened inventory
+    final List<FillerDefinition> fillers =
+        this.fillers.stream()
+            .filter(filler -> passesFilter.test(filler.getFilter()))
+            .collect(Collectors.toList());
+    if (fillers.isEmpty()) return;
+
+    logger.fine(
+        () ->
+            "Found fillers "
+                + fillers.stream()
+                    .map(FillerDefinition::toString)
+                    .collect(Collectors.joining(", ")));
+
+    // Find all Caches that the opened inventory is part of
+    final List<Fillable> fillables = new ArrayList<>();
+    for (Cache cache : caches) {
+      if (passesFilter.test(cache.region()) && passesFilter.test(cache.filter())) {
+        fillables.add(new FillableCache(cache));
+      }
+    }
+    // If the inventory is not in any Cache, just fill it directly
+    if (fillables.isEmpty()) {
+      fillables.add(new FillableInventory(inventory));
+    }
+
+    fillables.forEach(fillable -> fillable.fill(opener, fillers));
+  }
+
+  private boolean isFillerReady(Fillable fillable, FillerDefinition filler) {
+    if (filler.usesRefillInterval() || (!triggerFillerChanges && filler.usesRefillTrigger())) {
+      return filledAt.putUnlessNewer(
+              new FillableAndFiller(fillable, filler), filler.getRefillInterval())
+          == null;
+    }
+    final Integer count = this.triggerHistory.get(fillable, filler);
+    return triggeredFillers.containsKey(filler)
+        && (count == null || triggeredFillers.get(filler) > count);
+  }
+
+  abstract class Fillable {
+
+    abstract Stream<Inventory> inventories();
+
+    void fill(MatchPlayer opener, List<FillerDefinition> fillers) {
+      // Build a short list of Fillers that are NOT cooling down from a previous fill
+      final List<FillerDefinition> coolFillers =
+          fillers.stream()
+              .filter(filler -> isFillerReady(this, filler))
```

Comment:
Given isFillerReady is only used from within Fillable, maybe move it in so you don't need to pass `this` as first parameter?
```suggestion
              .filter(this::isFillerReady)
```
Would allow you to also use a method ref here instead of a lambda

Reply:
Sure, but i'm not sure how this affects size of each fillable and if its even relevant?

Reply:
a function does not affect memory size, functions are on the class not on the instance